File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/UseCase/UseCasesUseFactory.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class UseCasesUseFactory: QuickSpec {
    override class func spec() {
        Given("A UseCase") {
            let useCases = WoofHarmonize.useCases

            Then("It is annotated with @Factory") {
                useCases.assertTrue(rule: rule) { $0.hasAttribute(named: "@Factory") }
            }
        }
    }

    private static let rule = Rule(
        description: "UseCases must be annotated with @Factory.",
        rationale: "UseCases are stateless, so a new instance per injection is free and rules out accidental shared state.",
        fixHint: "Add @Factory to the UseCase class.",
        badExample: "@Single\nfinal class SendWoofUseCase",
        goodExample: "@Factory\nfinal class SendWoofUseCase"
    )
}

File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/usecase/UseCasesUseFactory.kt
package com.perrystreet.woof.konsist.usecase

import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class UseCasesUseFactory : BehaviorSpec() {
    init {
        Given("A UseCase") {
            val useCases = KonsistUtils.useCases

            Then("It is annotated with @Factory") {
                useCases.assertTrue(message = Message) { it.hasAnnotationWithName("Factory") }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "UseCases must be annotated with @Factory.",
            why = "UseCases are stateless, so a new instance per injection is free and rules out accidental shared state.",
            howToFix = "Add @Factory to the UseCase class.",
            badExample = "@Single\nclass SendWoofUseCase(...)",
            goodExample = "@Factory\nclass SendWoofUseCase(...)",
        )
    }
}
